### DISCUSSION TITLE
(🐞) Semver calculator acts differently to npm in regards to pre releases

### DISCUSSION THREAD

**Discussion by KotlinIsland** [community/community#128422] [npm]
- See https://github.com/npm/node-semver/issues/605

```json
"typescript": "^5 || ^5.3.0-0"
```
This value will match all stable versions of `5`, and all pre-releases of `5.3.0`.

<details><summary>Proof</summary>
<p>

```json
{
  "name": "test-npm",
  "version": "1.0.0",
  "lockfileVersion": 3,
  "requires": true,
  "packages": {
    "": {
      "name": "test-npm",
      "version": "1.0.0",
      "dependencies": {
        "typescript": "^5 || ^5.3.0-0"
      }
    },
    "node_modules/typescript": {
      "version": "5.3.0-dev.20230913",
      "resolved": "https://registry.npmjs.org/typescript/-/typescript-5.3.0-dev.20230913.tgz",
      "integrity": "sha512-+1lrLCj/NdyBbWXK3ifRSzmg0pSg2WLDmzeXf1/vTOb84eV/TCJ1i9eOS4c40T3tzDSAOjP39x9/+unUakbuKA==",
      "bin": {
        "tsc": "bin/tsc",
        "tsserver": "bin/tsserver"
      },
      "engines": {
        "node": ">=14.17"
      }
    }
  }
}
```

</p>
</details> 

But when we look on the website, it says that it won't match the pre-releases:

![image](https://github.com/npm/feedback/assets/65446343/6963546b-d9d9-4c78-a30f-ccd0c3b7f4bc)

According to [a comment](https://github.com/npm/node-semver/issues/605#issuecomment-1701946451) by @mbtools, this is intentional, but to me seems like it will only ever cause confusion.

**Comment by IanVS** [comment]
I think that might be because the website does not show pre-releases at all, ever, now (https://github.com/npm/feedback/discussions/995).

**Reply by mbtools** [reply]
Not entirely true. It includes pre-releases, if a condition includes pre-releases. 

![image](https://github.com/npm/feedback/assets/59966492/7abc49b7-fca5-4e04-afa8-fd5f6b00bc57)

![image](https://github.com/npm/feedback/assets/59966492/6a768132-30fa-41c6-a614-0efbe3e3c5da)

**Reply by IanVS** [reply]
Ah, ok.

I probably had checked a range like this (there are 5.0.0-beta.X): 

<img width="1227" alt="image" src="https://github.com/npm/feedback/assets/4616705/c6cf2291-cd68-4ba7-81ac-f9cbe52b296a">

Which is fine for peerDependencies, but acts differently for installing a package fresh, I guess.  It feels like the calculator website should have a "peerDependencies" mode, since that's usually when things get really tricky.
